hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -0,0 +1,1117 @@
+
+        return new IndexComponentCandidate(
+            exp,
+            columnIndex,
+            filter
+        );
+    }
+
+    private static IndexComponentCandidate prepareSingleColumnCandidateComparison(
+        RexNode exp,
+        SqlKind kind,
+        RexNode operand1,
+        RexNode operand2,
+        QueryParameterMetadata parameterMetadata
+    ) {
+        // Normalize operand positions, so that the column is always goes first
+        if (operand1.getKind() != SqlKind.INPUT_REF && operand2.getKind() == SqlKind.INPUT_REF) {
+            kind = inverseIndexConditionKind(kind);
+
+            RexNode tmp = operand1;
+            operand1 = operand2;
+            operand2 = tmp;
+        }
+
+        // Exit if the first operand is not a column
+        if (operand1.getKind() != SqlKind.INPUT_REF) {
+            return null;
+        }
+
+        int columnIndex = ((RexInputRef) operand1).getIndex();
+
+        if (!IndexRexVisitor.isValid(operand2)) {
+            // The second operand cannot be used for index filter
+            return null;
+        }
+
+        Expression<?> filterValue = convertToExpression(operand2, parameterMetadata);
+
+        if (filterValue == null) {
+            // Operand cannot be converted to expression. Do not throw an exception here, just do not use the faulty condition
+            // for index. The proper exception will be thrown on later stages when attempting to convert Calcite rel tree to
+            // Hazelcast plan.
+            return null;
+        }
+
+        IndexFilterValue filterValue0 = new IndexFilterValue(
+            singletonList(filterValue),
+            singletonList(false)
+        );
+
+        IndexFilter filter;
+
+        switch (kind) {
+            case EQUALS:
+                filter = new IndexEqualsFilter(filterValue0);
+
+                break;
+
+            case GREATER_THAN:
+                filter = new IndexRangeFilter(filterValue0, false, null, false);
+
+                break;
+
+            case GREATER_THAN_OR_EQUAL:
+                filter = new IndexRangeFilter(filterValue0, true, null, false);
+
+                break;
+
+            case LESS_THAN:
+                filter = new IndexRangeFilter(null, false, filterValue0, false);
+
+                break;
+
+            default:
+                assert kind == SqlKind.LESS_THAN_OR_EQUAL;
+
+                filter = new IndexRangeFilter(null, false, filterValue0, true);
+        }
+
+        return new IndexComponentCandidate(
+            exp,
+            columnIndex,
+            filter
+        );
+    }
+
+    private static Expression<?> convertToExpression(RexNode operand, QueryParameterMetadata parameterMetadata) {
+        try {
+            RexToExpressionVisitor visitor = new RexToExpressionVisitor(FieldTypeProvider.INSTANCE, parameterMetadata);
+
+            return operand.accept(visitor);
+        } catch (Exception e) {
+            return null;
```

Comment:
Here we try to convert Calcite expression to HZ expression. Normally it should never fail. But given that the Calcite is a kind of black box for us, we may get some unexpected behavior here. 

It is ok to ignore the exception at this stage because the same visitor will be invoked later after planning, and the exception will be re-thrown to the user with proper wrapping.